src/config.js

```diff
@@ -0,0 +1,107 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+const registryRegex = /(@[a-zA-Z1-9-]+:)?registry=https:(\/\/[a-zA-Z1-9-]+[-]npm[.]pkg[.]dev\/.*\/)/;
+const authTokenRegex = /(\/\/[a-zA-Z1-9-]+[-]npm[.]pkg[.]dev\/.*\/):_authToken=(.*)/;
+const passwordRegex = /(\/\/[a-zA-Z1-9-]+[-]npm[.]pkg[.]dev\/.*\/):_password=(.*)/;
+
+const configType = {
+  Default: "Default",
+  Registry: "Registry",
+  AuthToken: "AuthToken",
+  Password: "Password",
+}
+
+class Config {
+  constructor(text) {
+    this.text = text;
+    this.type = configType.Default;
+  }
+
+  toString() {
+    return this.text;
+  }
+}
+
+
+class RegistryConfig {
+  constructor(scope, registry) {
+    this.scope = scope;
+    this.registry = registry;
+    this.type = configType.Registry;
+  }
+
+  toString() {
+    let prefix = this.scope ? this.scope + ':' : '';
+    return `${prefix}registry=https:${this.registry}`;
+  }
+}
+
+class AuthTokenConfig {
+  constructor(registry, token) {
+    this.registry = registry;
+    this.token = token;
+    this.type = configType.AuthToken;
+  }
+
+  toString() {
+    return `${this.registry}:_authToken=${this.token}`;
+  }
+
+  refreshToken(token) {
+    this.token = token;
+  }
+}
+
+class PasswordConfig {
+  constructor(registry, password) {
+    this.registry = registry;
+    this.password = password;
+    this.type = configType.Password;
+  }
+
+  toString() {
+    return `${this.registry}:_password=${this.password}`;
+  }
+}
+
+function parseConfig(text) {
```

Comment:
the implementation of this function (and most everything in this file) does not seem well aligned to the idioms of javascript. Do you really need all the concrete "Config" classes? Why not just return a simple object literal with the `toString` value inline:

```javascript
export const parseConfig = (text) => {
  let m = text.match(registryRegex);
  if (m) {
    return {
      configType: ConfigType.Registry,
      sope: m[1] ? m[1].replace(':', '') : m[1],
      registry: m[2],
      configValue: `${m[1]}registry=https:${this.registry}`;
    };
  }
  // other ones... 
  return {
    configType: ConfigType.Default,
    configValue: text
  };
}
```
I think this implementation will simplify a lot of the code that calls this method (eg the unit tests).

Reply:
Done. Had to make `configValue` a function, so that when update the `token` field `configValue` is updated too.